### PR TITLE
PTRENG-4916 framework migration

### PR DESCRIPTION
**Pull request by danielmkn**
Start of migration from SDK v2 to Terraform Plugin Framework.
- provider muxing;
- migrated `artifactory_user`, `artifactory_anonymous_user` and `artifactory_managed_user` resources;
- replaced terraform-plugin-sdk/v2/ package with terraform-plugin-testing in all go files;
- renamed `test` package in shared module to `testutil` to avoid conflicts in test files with Go `test` package;
- added templates and examples for auto-generated documentation for users;

### CODE REVIEW

File: docs/resources/managed_user.md [hunk 1]
```diff
@@ -1,44 +1,55 @@
...
-Consider using a separate provider to generate and manage passwords. 
+Consider using a separate provider to generate and manage passwords.
 
 ~> The password is stored in the Terraform state file. Make sure you secure it, please refer to the official [Terraform documentation](https://developer.hashicorp.com/terraform/language/state/sensitive-data).
 
 ## Example Usage
 
-```hcl
-# Create a new Artifactory user called terraform
+```terraform
```
Comment (chb0github) [inline docs/resources/managed_user.md:17]: I've seen GH handle both of these declarations. I don't have a dog in this fight, but I'm curious why the change. Was `hcl` not working?

File: docs/resources/managed_user.md [hunk 2]
```diff
@@ -1,44 +1,55 @@
...
 
-* `name` - (Required) Username for user.
-* `email` - (Required) Email for user.
-* `password` - (Required) Password for the user.
-* `admin` - (Optional) When enabled, this user is an administrator with all the ensuing privileges. Default value is `false`.
-* `profile_updatable` - (Optional) When set, this user can update his profile details (except for the password. Only an administrator can update the password). Default value is `true`.
-* `disable_ui_access` - (Optional) When set, this user can only access Artifactory through the REST API. This option cannot be set if the user has Admin privileges. Default value is `true`.
-* `internal_password_disabled` - (Optional) When set, disables the fallback of using an internal password when external authentication (such as LDAP) is enabled.
-* `groups` - (Optional) List of groups this user is a part of. **Notes:** If this attribute is not specified then user's group membership is set to empty. User will not be part of default "readers" group automatically.
+### Optional
```
Comment (chb0github) [inline docs/resources/managed_user.md:36]: This is all generated? Interesting

File: pkg/acctest/test.go [hunk 3]
```diff
@@ -11,13 +11,14 @@ import (
 
 	"github.com/go-resty/resty/v2"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	terraform2 "github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
```
Comment (chb0github) [inline pkg/acctest/test.go:14]: I don't get why you have the same import twice, one aliased

File: docs/resources/managed_user.md (hunk 2 above)
Comment (danielmkn) [reply docs/resources/managed_user.md:36]: Yes.

File: pkg/acctest/test.go (hunk 3 above)
Comment (danielmkn) [reply pkg/acctest/test.go:14]: It's not the same import. 
We need an alias because of this conversion:
```
configErr := Provider.Configure(context.Background(), (*terraform2.ResourceConfig)(terraform.NewResourceConfigRaw(nil)))
```

File: pkg/artifactory/resource/user/resource_artifactory_user_test.go [hunk 4]
```diff
@@ -5,182 +5,234 @@ import (
...
 )
 
-func TestAccUserPasswordNotChangeWhenOtherAttributesChangeGH340(t *testing.T) {
-	id := test.RandomInt()
-	name := fmt.Sprintf("user-%d", id)
-	fqrn := fmt.Sprintf("artifactory_user.%s", name)
+var (
+	// these will be set by the goreleaser configuration
+	// to appropriate values for the compiled binary
+	version = "test"
```
Comment (alexhung) [inline pkg/artifactory/resource/user/resource_artifactory_user_test.go:23]: This won't be necessary once we change how to set the provider's version string for framework.

**Review by alexhung — CHANGES_REQUESTED** [review]

File: pkg/artifactory/datasource/security/datasource_artifactory_group.go [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package security
 
 import (
+	utilsdk "github.com/jfrog/terraform-provider-shared/util/sdk"
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/datasource/security/datasource_artifactory_group.go:4]: find/replace stuff, moved.

File: pkg/artifactory/provider/provider_fw.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+			},
+			"access_token": schema.StringAttribute{
+				MarkdownDescription: "This is a access token that can be given to you by your admin under `Identity and Access`. If not set, the 'api_key' attribute value will be used.",
+				Optional:            true,
+				Sensitive:           true,
+			},
+			"check_license": schema.BoolAttribute{
+				MarkdownDescription: "Toggle for pre-flight checking of Artifactory Pro and Enterprise license. Default to `true`.",
+				Optional:            true,
+			},
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/provider/provider_fw.go:58]: There is no default in the resource schema in the framework.

File: main.go [hunk 7]
```diff
@@ -1,12 +1,53 @@
...
-	})
+	ctx := context.Background()
+
+	var debug bool
+
+	flag.BoolVar(&debug, "debug", false, "set to true to run the provider with support for debuggers like delve")
+	flag.Parse()
+
+	providers := []func() tfprotov5.ProviderServer{
+		providerserver.NewProtocol5(provider.New(provider.Version)()), // Example terraform-plugin-framework provider
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply main.go:28]: Good idea, changed.

File: pkg/artifactory/provider/provider_fw.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+		},
+	}
+}
+
+func (p *ArtifactoryProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
+	// Provider specific implementation.
+
+	// Check environment variables first
+	url := os.Getenv("JFROG_URL")
+	accessToken := os.Getenv("JFROG_ACCESS_TOKEN")
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/provider/provider_fw.go:68]: Added.

File: pkg/artifactory/provider/provider_fw.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+func (p *ArtifactoryProvider) DataSources(ctx context.Context) []func() datasource.DataSource {
+	return []func() datasource.DataSource{
+		// Provider specific implementation
+	}
+}
+
+func New(version string) func() provider.Provider {
+	return func() provider.Provider {
+		return &ArtifactoryProvider{
+			version: version,
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/provider/provider_fw.go:175]: Changed the version approach entirely now.

File: pkg/artifactory/provider/provider_fw_test.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+					return nil, err
+				}
+
+				return muxServer.ProviderServer(), nil
+			},
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(providerConfig, url, token),
+			},
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/provider/provider_fw_test.go:48]: At least nothing in the HashiCorp documentation. I rely on the absence of errors in `providers` in row 31.

File: pkg/artifactory/resource/user/resource_artifactory_user_test.go (hunk 4 above)
Comment (danielmkn) [reply pkg/artifactory/resource/user/resource_artifactory_user_test.go:23]: Removed from all tests.

Comment (danielmkn) [reply pkg/artifactory/resource/user/resource_artifactory_user_test.go:23]: I'm setting the test version in the go run parameters now.

File: pkg/artifactory/resource/user/resource_artifactory_managed_user.go [hunk 11]
```diff
@@ -3,40 +3,40 @@ package user
...
+}
 
-		Description: "Provides an Artifactory managed user resource. This can be used to create and manage Artifactory users. For example, service account where password is known and managed externally.",
+func (r *ArtifactoryManagedUserResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	var managedUserSchemaFramework = map[string]schema.Attribute{
+		"password": schema.StringAttribute{
+			MarkdownDescription: "Password for the user. Must be 12 characters with 1 digit, 1 symbol, with upper and lower case letters.",
+			Required:            true,
+			Sensitive:           true,
+			Validators:          []validator.String{stringvalidator.LengthAtLeast(12)},
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply pkg/artifactory/resource/user/resource_artifactory_managed_user.go:32]: Added regex.

File: pkg/acctest/test.go [hunk 12]
```diff
@@ -62,15 +63,15 @@ func PreCheck(t *testing.T) {
...
+		configErr := Provider.Configure(context.Background(), (*terraform2.ResourceConfig)(terraform.NewResourceConfigRaw(nil)))
 		if configErr != nil {
 			t.Fatalf("Failed to configure provider %v", configErr)
 		}
 	})
 }
 
 func GetArtifactoryUrl(t *testing.T) string {
-	return test.GetEnvVarWithFallback(t, "ARTIFACTORY_URL", "JFROG_URL")
+	return testutil.GetEnvVarWithFallback(t, "ARTIFACTORY_URL", "JFROG_URL")
```
Comment (alexhung) [inline pkg/acctest/test.go:74]: We should swap the ordering of these 2 env vars as well

File: pkg/artifactory/resource/user/resource_artifactory_user_test.go [hunk 13]
```diff
@@ -5,182 +5,228 @@ import (
...
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
 	"github.com/jfrog/terraform-provider-artifactory/v7/pkg/acctest"
-	"github.com/jfrog/terraform-provider-shared/test"
-	"github.com/jfrog/terraform-provider-shared/util"
+	"github.com/jfrog/terraform-provider-artifactory/v7/pkg/artifactory/provider"
+	"github.com/jfrog/terraform-provider-shared/testutil"
+	utilsdk "github.com/jfrog/terraform-provider-shared/util/sdk"
 	"github.com/jfrog/terraform-provider-shared/validator"
+	//"github.com/mitchellh/go-testing-interface"
```
Comment (alexhung) [inline pkg/artifactory/resource/user/resource_artifactory_user_test.go:17]: Still need this?

File: pkg/acctest/test.go (hunk 12 above)
Comment (danielmkn) [reply pkg/acctest/test.go:74]: yes.

File: pkg/artifactory/resource/user/resource_artifactory_user_test.go (hunk 13 above)
Comment (danielmkn) [reply pkg/artifactory/resource/user/resource_artifactory_user_test.go:17]: Removed.

**Review by alexhung — APPROVED** [review]
